cyral/resource_cyral_sidecar_listener.go

```diff
@@ -0,0 +1,363 @@
+	DynamoDbSettingsKey = "dynamodb_settings"
+)
+
+// SidecarListener struct for sidecar listener.
+type SidecarListener struct {
+	SidecarId        string            `json:"-"`
+	ListenerId       string            `json:"id"`
+	RepoTypes        []string          `json:"repoTypes"`
+	NetworkAddress   *NetworkAddress   `json:"address,omitempty"`
+	MySQLSettings    *MySQLSettings    `json:"mysqlSettings,omitempty"`
+	S3Settings       *S3Settings       `json:"s3Settings,omitempty"`
+	DynamoDbSettings *DynamoDbSettings `json:"dynamoDbSettings,omitempty"`
+}
+type NetworkAddress struct {
+	Host string `json:"host,omitempty"`
+	Port int    `json:"port"`
+}
+type MySQLSettings struct {
+	DbVersion    string `json:"dbVersion,omitempty"`
+	CharacterSet string `json:"characterSet,omitempty"`
+}
+type S3Settings struct {
+	ProxyMode bool `json:"proxyMode,omitempty"`
+}
+type DynamoDbSettings struct {
+	ProxyMode bool `json:"proxyMode,omitempty"`
+}
+
+var ReadSidecarListenersConfig = ResourceOperationConfig{
+	Name:       "SidecarListenersResourceRead",
+	HttpMethod: http.MethodGet,
+	CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+		return fmt.Sprintf("https://%s/v1/sidecars/%s/listeners/%s",
+			c.ControlPlane,
+			d.Get(SidecarIdKey).(string),
+			d.Get(ListenerIdKey).(string))
+	},
+	NewResponseData: func(_ *schema.ResourceData) ResponseData { return &ReadSidecarListenersAPIResponse{} },
+}
+
+type ReadSidecarListenersAPIResponse struct {
+	ListenerConfig *SidecarListener `json:"listenerConfig"`
+}
+type CreateListenerAPIResponse struct {
+	ListenerId string `json:"listenerId"`
+}
+
+func (c CreateListenerAPIResponse) WriteToSchema(d *schema.ResourceData) error {
+	d.SetId(marshalComposedID([]string{d.Get(SidecarIdKey).(string), c.ListenerId}, "/"))
+	return d.Set(ListenerIdKey, c.ListenerId)
+}
+
+func (data ReadSidecarListenersAPIResponse) WriteToSchema(d *schema.ResourceData) error {
+	if data.ListenerConfig != nil {
+		_ = d.Set(ListenerIdKey, data.ListenerConfig.ListenerId)
+		_ = d.Set(RepoTypesKey, data.ListenerConfig.RepoTypesAsInterface())
+		_ = d.Set(NetworkAddressKey, data.ListenerConfig.NetworkAddressAsInterface())
+		_ = d.Set(S3SettingsKey, data.ListenerConfig.S3SettingsAsInterface())
+		_ = d.Set(MySQLSettingsKey, data.ListenerConfig.MySQLSettingsAsInterface())
+		_ = d.Set(DynamoDbSettingsKey, data.ListenerConfig.DynamoDbSettingsAsInterface())
+	}
+	return nil
+}
+func (l *SidecarListener) RepoTypesAsInterface() []interface{} {
+	if l.RepoTypes == nil {
+		return nil
+	}
+	result := make([]interface{}, len(l.RepoTypes))
+	for i, v := range l.RepoTypes {
+		result[i] = v
+	}
+	return result
+}
+func (l *SidecarListener) RepoTypesFromInterface(anInterface []interface{}) {
+	repoTypes := make([]string, len(anInterface))
+	for i, v := range anInterface {
+		repoTypes[i] = v.(string)
+	}
+	l.RepoTypes = repoTypes
+}
+func (l *SidecarListener) NetworkAddressAsInterface() []interface{} {
+	if l.NetworkAddress != nil {
```

Comment:
Nit: We could make the nil checks for all the *AsInterface methods more consistent by using either:
```suggestion
	if l.NetworkAddress == nil {
		return nil
	}
```
or
```suggestion
	if l.NetworkAddress != nil {
		...
	}
	return nil
```